### PR TITLE
Shader Lab support `RenderState`

### PR DESCRIPTION
**Pull request by Sway007**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
- `RenderState` can be set as follows in ShaderLab

```shader
Shader "demo" {
  SubShader {
    Pass "default" {
      ...
      StencilState {
         Enabled = true;
         ReferenceValue = xxx;
         Mask = xx.xx; // 0xffffffff
         WriteMask = xx.xx; // 0xffffffff
         CompareFunctionFront = CompareFunction.Less;
      }
      ...
    }
  }
}
```
- Imporoved support of macro syntax

### CODE REVIEW

File: packages/shader-lab/src/RuntimeContext.ts [hunk 1]
```diff
@@ -90,99 +111,178 @@ export default class RuntimeContext {
...
+        ast: struct,
+        name: struct.content.name
+      })) ?? []),
+      ...(this.shaderAst.content.renderStates?.map((v) => ({ referenced: false, ast: v, name: v.content.variable })) ??
+        []),
+      ...(this.shaderAst.content.variables?.map((item) => ({
+        referenced: false,
+        ast: item,
+        name: item.content.variable
+      })) ?? [])
```
Comment (GuoLei1990) [inline packages/shader-lab/src/RuntimeContext.ts:128]: Just want add property `referenced`?

File: packages/shader-lab/scripts/GenDts.ts [hunk 2]
```diff
@@ -6,7 +6,13 @@ import fs from "fs";
 function genDts(opts?: { outDir?: string }) {
   const parser = new ShaderParser();
   const productions = parser.getGAstProductions();
-  const dtsContent = generateCstDts(productions);
+  let dtsContent = generateCstDts(productions);
+
+  dtsContent = dtsContent.replace(
+    /((BlendFactor|BlendOperation|CompareFunction|StencilOperation|CullMode)\.\w+?)\?:/g,
+    '"$1"?:'
+  );
```
Comment (GuoLei1990) [inline packages/shader-lab/scripts/GenDts.ts:14]: something hack?

Comment (Sway007) [reply packages/shader-lab/scripts/GenDts.ts:14]: Yeah, the type generation tool of chevrotain does not fully support the declaration of complex type properties. The code above will not run at runtime

File: packages/shader-lab/src/RuntimeContext.ts (hunk 1 above)
Comment (Sway007) [reply packages/shader-lab/src/RuntimeContext.ts]: no, have been explained on the phone :)

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**